tests/unit/test_utils.py

```diff
@@ -84,41 +90,83 @@ def test_run_cmd_executable_not_found(
         run_cmd(["foo"], params={})
 
 
-@mock.patch("shutil.copytree")
-@mock.patch("reflink.supported_at")
+@mock.patch("cachi2.core.utils._get_blocksize")
+def test_fast_copy(mock_blocksize: mock.Mock, tmp_path: Path) -> None:
+    mock_blocksize.return_value = 4
+    src = tmp_path / "src/test"
+    dest = tmp_path / "dest/test"
+    for dir_ in [src.parent, dest.parent]:
+        dir_.mkdir()
+    test_str = "hello world"
+
+    with open(src, "w") as fd:
+        fd.write(test_str)
+
+    nbytes = _fast_copy(src, dest)
+    assert nbytes == len(test_str)
+
+
 @pytest.mark.parametrize(
-    "reflink_supported",
-    (True, False),
+    "errno_",
+    [
+        pytest.param(errno.ENOSYS, id="ENOSYS"),
+        pytest.param(errno.EXDEV, id="EXDEV"),
+        pytest.param(errno.EAGAIN, id="unexpected_errno"),
+    ],
 )
-def test_copy_directory(
-    mock_reflink_supported_at: mock.Mock, mock_shutil_copytree: mock.Mock, reflink_supported: bool
-) -> None:
-    mock_reflink_supported_at.return_value = reflink_supported
+@mock.patch("os.copy_file_range")
+def test_fast_copy_fail_errno(mock_copy_range: mock.Mock, tmp_path: Path, errno_: int) -> None:
+    src = tmp_path / "src/test"
+    dest = tmp_path / "dest/test"
+    for dir_ in [src.parent, dest.parent]:
+        dir_.mkdir()
 
-    origin = Path("/fake")
-    destination = Path("/phony")
+    mock_copy_range.side_effect = OSError(errno_)
 
-    copy_directory(origin, destination)
+    with pytest.raises(Exception) as ex:
+        _fast_copy(src, dest)
+
+        if errno_ == errno.EAGAIN:
+            isinstance(ex, OSError)
+        else:
+            isinstance(ex, _FastCopyFailedFallback)
 
-    if reflink_supported:
-        copy_function = reflink.reflink
-    else:
-        copy_function = shutil.copy2
 
-    assert mock_shutil_copytree.call_args.kwargs["copy_function"] == copy_function
+@mock.patch("cachi2.core.utils.open")
+def test_fast_copy_fail_io_fileno(mock_open: mock.MagicMock, tmp_path: Path) -> None:
+    """Test that we correctly fall back to regular copy when encountering a non-regular file."""
```

Comment:
This docstring is a bit misleading (the one on L152 too): here we do not test that a fallback is happening, but rather that the need to fallback was signaled.